Libplanet/Net/Swarm.MessageHandlers.cs

```diff
@@ -212,25 +210,28 @@ private void ProcessTxIds(TxIds message)
                 message.Ids.Select(txid => txid.ToString())
             );
 
-            IStagePolicy<T> stagePolicy = BlockChain.StagePolicy;
-            ImmutableHashSet<TxId> newTxIds = message.Ids
-                .Where(id => !_demandTxIds.ContainsKey(id))
-                .Where(id => !stagePolicy.Ignores(BlockChain, id))
-                .ToImmutableHashSet();
+            HashSet<TxId> required = GetRequiredTxIds(message.Ids);
 
-            if (!newTxIds.Any())
+            if (!required.Any())
             {
                 _logger.Debug("No unaware transactions to receive.");
                 return;
             }
 
             _logger.Debug(
                 "Unaware transactions to receive: {@TxIds}.",
-                newTxIds.Select(txid => txid.ToString())
+                required.Select(txid => txid.ToString())
             );
-            foreach (TxId txid in newTxIds)
+
+            if (_txSyncTaskQueue.ContainsKey(peer))
+            {
+                _txSyncTaskQueue[peer].UnionWith(required);
```

Comment:
AFIK, this code can be access from multiple task(thread)s and `HashSet<T>` isn't thread-safe. so I think we may need to allow concurrent download for the same tx id from multiple threads and coordinate them.